routers/private/hook.go

```diff
@@ -92,6 +105,154 @@ func HookPreReceive(ctx *macaron.Context, opts private.HookOptions) {
 
 				}
 			}
+
+			// Require signed commits
+			if protectBranch.RequireSignedCommits {
+				env := os.Environ()
+				if opts.GitAlternativeObjectDirectories != "" {
+					env = append(env,
+						private.GitAlternativeObjectDirectories+"="+opts.GitAlternativeObjectDirectories)
+				}
+				if opts.GitObjectDirectory != "" {
+					env = append(env,
+						private.GitObjectDirectory+"="+opts.GitObjectDirectory)
+				}
+				if opts.GitQuarantinePath != "" {
+					env = append(env,
+						private.GitQuarantinePath+"="+opts.GitQuarantinePath)
+				}
+
+				stdoutReader, stdoutWriter, err := os.Pipe()
+				if err != nil {
+					log.Error("Unable to create pipe: %v", err)
+					ctx.JSON(http.StatusInternalServerError, map[string]interface{}{
+						"err": fmt.Sprintf("Unable to create pipe: %v", err),
+					})
+					return
+				}
+				defer func() {
+					_ = stdoutReader.Close()
+					_ = stdoutWriter.Close()
+				}()
+
+				stderr := new(bytes.Buffer)
+				commits := make([]*git.Commit, 0, 10)
+
+				var finalErr error
+				if err := git.NewCommand("rev-list", oldCommitID+"..."+newCommitID).
```

Comment:
move this whole routine in a seperate function located at `services` or some other git-fuu-related place ...?